chore: updated contributing doc (#1625)
instana/nodejs: CONTRIBUTING.md

We have added a CI build to test our instrumentations against ES module apps.
See https://github.com/instana/nodejs/pull/672

Not all of the current test apps have an `.mjs` variant, because the effort is high. If you are adding a new test, please consider to also generate an `.mjs` file for it. If you are modifying an existing application under test, and it has an accompanying `.mjs` file, you need to update that as well (by regenerating it).

You can use the script `bin/convert-test-app-to-es6.sh` for both purposes, it generates an equivalent `.mjs` file from a given `.js` file. 

For example, run `bin/convert-test-app-to-es6.sh packages/collector/test/tracing/database/ioredis/app.js` to regenerate the ES6 variant of the `ioredis` test application, when you have made changes to `packages/collector/test/tracing/database/ioredis/app.js`.

After regenerating an existing `.mjs` file you should check the diff for it and potentially revert any changes that are artifacts of the conversion process, not related to your original changes in `.js` file.

> If your changes introduce ESM files other than `app.mjs`, the ESM execution logic requires you to create `.mjs` files for all applications within the same package.

Reviewer: > introduce ESM files

Its not clear from this sentence, that this is about multiple ESM apps in one folder.

For example, if a directory contains the following files (all these are different applications with with different usecases):

 - sender.js
 - app.js
 - rootExitSpanApp.js

and you add a test for `sender.mjs`, you must also ensure that `.mjs` versions exist for `app.js` and `rootExitSpanApp.js`. Failing to do so may cause issues when running ESM tests.

> 

Reviewer: suggestion: Add a description for __dirname vs __appPath? This is missing.

We have added a CI build to test our instrumentations against ES module apps.
See https://github.com/instana/nodejs/pull/672

Not all of the current test apps have an `.mjs` variant, because the effort is high. If you are adding a new test, please consider to also generate an `.mjs` file for it. If you are modifying an existing application under test, and it has an accompanying `.mjs` file, you need to update that as well (by regenerating it).

You can use the script `bin/convert-test-app-to-es6.sh` for both purposes, it generates an equivalent `.mjs` file from a given `.js` file. 

For example, run `bin/convert-test-app-to-es6.sh packages/collector/test/tracing/database/ioredis/app.js` to regenerate the ES6 variant of the `ioredis` test application, when you have made changes to `packages/collector/test/tracing/database/ioredis/app.js`.

After regenerating an existing `.mjs` file you should check the diff for it and potentially revert any changes that are artifacts of the conversion process, not related to your original changes in `.js` file.

> If you add an ESM app to an existing directory, the ESM execution logic requires `.mjs` versions of all existing applications within that directory.

Reviewer: ```suggestion
If you add ESM support to a test directory, you have to transform **all** existing CJS apps to ESM apps.
```

> If your changes introduce multiple ESM files, the ESM execution logic requires `.mjs` versions for all existing applications within the same directory.

Reviewer 2: ```suggestion
If you add an ESM app to an existing directory, the ESM execution logic requires `.mjs` versions of all existing applications within that directory.
```

Example: Consider a directory containing the following files, each representing a different application with a unique use case:

 - sender.js
 - app.js
 - rootExitSpanApp.js
 

> If you add a test for `sender.mjs`, you must also ensure that `.mjs` versions exist for `app.js` and `rootExitSpanApp.js`. Failing to do so may cause issues when running ESM tests.

Reviewer 2: ```suggestion
If you add an ES module application for `sender.js`, you must also ensure that equivalent `.mjs` ESM applications exist for `app.js` and `rootExitSpanApp.js`. Otherwise, it may lead to issues when running ESM tests.
```

> If you add an ES module application for `sender.js`, you must also ensure that equivalent `.mjs` ESM applications exist for `app.js` and `rootExitSpanApp.js`. Otherwise, it may lead to issues when running ESM tests.

Reviewer: > it may lead to issues when running ESM tests.

Which issue? I am missing the link here.

Author: We could say issues in general because we were getting timing issues with ESM in messaging tests. 
allowRootExitSpanApp was running for ESM even if there is no `.mjs` file.

We can't limit the execution partially in a test suit if we have multiple `describe` statements

Reviewer: But can't we simply link the the fixes we tried?

Reviewer: We need a reminder here what the issues were

You can use the script `bin/convert-test-app-to-es6.sh` for both purposes, it generates an equivalent `.mjs` file from a given `.js` file. 

For example, run `bin/convert-test-app-to-es6.sh packages/collector/test/tracing/database/ioredis/app.js` to regenerate the ES6 variant of the `ioredis` test application, when you have made changes to `packages/collector/test/tracing/database/ioredis/app.js`.

After regenerating an existing `.mjs` file you should check the diff for it and potentially revert any changes that are artifacts of the conversion process, not related to your original changes in `.js` file.

If you add ESM support to a test directory, you have to transform **all** existing CJS apps to ESM apps.

Example: Consider a directory containing the following files, each representing a different application with a unique use case:

 - sender.js
 - app.js
 - rootExitSpanApp.js
 
If you add an ESM application for `sender.js`, you must also ensure that equivalent `.mjs` ESM applications exist for `app.js` and `rootExitSpanApp.js`.

> Suppose in this case, if `rootExitSpanApp.mjs` is not added, the ESM test execution will default to running `rootExitSpanApp.js` alongside ESM unless the test is disabled for ESM. This can lead to timing issues due to differences in module loading between `.js` and `.mjs` files, potentially causing unintended span data.

Reviewer: I am still missing the Github links or Jira :D

Author: Added GitHub PR examples for the same 👍

Reviewer: Thanks<3
